### DISCUSSION TITLE
Feature Request: API

### DISCUSSION THREAD

**Discussion by pvieira84** [oss-apps/split-pro#300] [Ideas]
### Discussed in https://github.com/oss-apps/split-pro/discussions/88

<div type='discussions-op-text'>

<sup>Originally posted by **pvieira84** August 30, 2024</sup>
Any chance an API can be added to the roadmap? I believe it would be helpful and handy for a large group of people</div>

**Comment by Egglestron** [comment]
The only thing that would convince me to abandon Tricount!

**Comment by idoodler** [comment]
Would be awesome to integrate it with iOS Shortcuts

**Comment by krokosik** [comment]
We could use this maybe? https://github.com/mcampa/trpc-to-openapi

**Comment by krokosik** [comment]
@pvieira84 @idoodler @Egglestron I am working on an integration right now, which would enable a RESTful API together with a new OpenAPI json endpoint and a SwaggerUI page. Once that is done, would you be willing to help with writing description for the existing endpoints?

**Comment by krokosik** [comment]
I would like to further discuss what you would expect from a REST API. Does it have to cover ALL the endpoints or will a subset be enough? I am also not sure how credentials could be handled within NextAuth. I assume something like API keys would be preferred, but I don't see this functionality in the library.

The trpc-to-openapi would require extensive refactoring and testing, since it requires full input and output validation, which was sadly not implemented by the original author and it breaks some routes.

**Reply by Timmy93** [reply]
Hi krokosik, I don't know if it is too late, but for sure it would be fantastic to have at least the possibility to have read access to the expenses via api. 

For me a use case is the integration with firefly-iii. At the moment I'm stuck with Splitwise due to their api. 

Thank you for your fantastic job

**Reply by J4nsen** [reply]
For me a subset would be enough. atm I want to let my openclaw bot add new expenses

**Reply by krokosik** [reply]
As of late I have very limited time to work on SplitPro and an API is a tedious undertaking. I will keep in mind that there is renewed interest in this and maybe start adding routes incrementally. Could you list your must have endpoints? I will expose them first.

**Reply by J4nsen** [reply]
Hi krokosik, in my case thats probably

*  `expense.addOrEditExpense`
* `expense.getGroupExpenses`
* `expense.getExpenseDetails`
* `group.getAllGroups`
* `group.getGroupDetails`
* `user.me`

Thanks in advance, it is an awesome project :)

**Comment by ZenoBell** [comment]
Apologies, I  just noticed too late now that there was this thread already opened and I opened a separate one. 

https://github.com/oss-apps/split-pro/discussions/641 

Instead of using somethign like @krokosik was suggesting like trpc-to-openapi , I completely bypased tRPC and asked Claude to write me a brand new Next.js API that routes from scratch. It only covers the specific endpoints I needed (summary, groups, add/delete expense) . It uses the internal BalanceView directly via Prisma. So like that no refactoring was needed, and it works immediately. But yeah it's a parallel API sitting next to the app, not integrated with the existing tRPC layer.  

I'm aware this is far from a proper implementation ,  ideally it would use something like trpc-to-openapi to expose the existing tRPC routes as a standard REST API with OpenAPI spec and Swagger UI. But as the maintainer noted, that requires extensive refactoring and full input/output validation across all routes. This was my pragmatic workaround while waiting for that  maybe useful as a reference or to unblock people who need something working today.

Cheers
